File: net/lib/schemas/BaseSchema.cs
using System.Text.Json;
using Albert.Lib.Rules;

namespace Albert.Lib.Schemas;

public abstract class BaseSchema<T>
{
    private static readonly JsonSerializerOptions _jsonOptions = new() { WriteIndented = true };
    public abstract string Type { get; }
    private readonly List<IBaseRule<T>> _rules = [];


    public BaseSchema<T> AddRule(IBaseRule<T> rule)
    {
        _rules.Add(rule);
        return this;
    }

    public List<ValidationError> Validate(T? value)
    {
        var errors = new List<ValidationError>();
        foreach (var rule in _rules)
        {
            if (!rule.Validate(value))
            {
                // TODO: how to get a path?
                errors.Add(new ValidationError(rule.ErrorMessage(), null));
            }
        }

        return errors;
    }

    public void ValidateAndLog(T? value)
    {
        var errors = Validate(value);
        Console.WriteLine(JsonSerializer.Serialize(errors, _jsonOptions));
    }
}

File: js/lib/schemas/BaseSchema.ts
import { BaseRule } from '../rules';
import { ValidationError } from '../validation';

export abstract class BaseSchema<T> {
  __type!: T;
  type: string;
  rules: BaseRule<T>[] = [];

  constructor(type: string) {
    this.type = type;
  }

  addRule(rule: any) {
    this.rules.push(rule);
    return this;
  }

  validate(value?: T, path?: string) {
    const errors: ValidationError[] = [];
    for (const rule of this.rules) {
      if (!rule.validate(value)) {
        // TODO: how to get a path?
        errors.push(new ValidationError(rule.errorMessage(), path));
      }
    }

    return errors;
  }

  validateAndLog(value?: T) {
    const errors = this.validate(value);
    console.dir(errors);
  }
}

export type InferType<T extends BaseSchema<any>> = T['__type'];
